addons/mass_mailing/models/mailing.py

```diff
@@ -1335,6 +1342,52 @@ def _image_to_url(b64image: bytes):
                 node.attrib['src'] = _image_to_url(image)
                 modified = True
 
+    def _crop_vml_images(self, body_html):
+        """
+        Find VML v:image elements, crop their source images, make an attachement
+        out of them and replace their source with an url to the attachement.
+        """
+
+        def _image_to_url(b64image: bytes):
+            """Store an image in an attachement and returns an url"""
+            attachment = self.env['ir.attachment'].create({
+                'datas': b64image,
+                'name': "cropped_image_mailing_{}".format(self.id),
+                'type': 'binary',})
+
+            attachment.generate_access_token()
+
+            return '/web/image/%s?access_token=%s' % (
+                attachment.id, attachment.access_token)
+
+        root = lxml.html.fromstring(body_html)
+        modified = False
+        for comment in root.xpath('//comment()'):
+            is_mso = mso_re.match(comment.text)
+            if is_mso:
+                matches = re.findall(r'<v:image[^>]*>', comment.text)
+                if matches:
+                    for match in matches:
+                        try:
+                            url = re.search(r'src=\s*\"([^\"]+)\"', match).group(1)
+                            absolute_url = url
+                            if url.startswith('/'):
+                                absolute_url = self.get_base_url() + url
+                            urllib.request.urlretrieve(absolute_url, 'temp.jpg')
+                            buffered = BytesIO()
+                            image = Image.open('temp.jpg')
+                            image.save(buffered, format="JPEG")
+                            width = float(re.search(r'width:\s*([0-9]+)\s*px', match).group(1))
+                            height = float(re.search(r'height:\s*([0-9]+)\s*px', match).group(1))
+                            image_processor = tools.ImageProcess(buffered.getvalue())
+                            image = image_processor.crop_resize(width, height, 0, 0).image
+                            buffered = BytesIO()
+                            image.save(buffered, format="JPEG")
+                            comment.text = comment.text.replace(url, _image_to_url(base64.b64encode(buffered.getvalue())))
+                            modified = True
+                        except Exception:
+                            pass
```

Comment:
Wow, what a catchall. What kind of exceptions do you expect to get here ? Because here if you ever have an actual bug in the code, you'll never know it until a client sees it in its resulting email and you spend time to investigate the issue to realize the issue is actually hidden by this very broad except clause.

Reply:
You should know what kind of exception you can rightfully get and only catch those ones.

Reply:
Well done, thanks to this I actually caught an error in my regexes :cold_sweat:
Before: `width:\s*([0-9]+)\s*px`
After: `width:\s*([0-9\.]+)\s*px`

Here's my suggested change for the `except`:
```py
except (ValueError, UnidentifiedImageError, AttributeError):
    # - ValueError is raised when the url is not a valid url
    # - UnidentifiedImageError is raised when the url is not
    #   an image
    # - AttributeError is raised when one of the regexes
    #   fails to match (e.g. when the image has no width or
    #   height).
    # We choose to ignore these errors so as not to
    # break the entire process just for one image's
    # responsive cropping behavior).
    pass
```